src/8a860_len_3f30.c

```diff
@@ -88,7 +89,68 @@ void func_800F4C6C(PopupMessage* popup, s32 x, s32 y);
 void func_800F4CF0(PopupMessage* popup, s32 x, s32 y);
 void func_800F4D28(PopupMessage* popup, s32 x, s32 y);
 
-INCLUDE_ASM(s32, "8a860_len_3f30", hide_popup_menu);
+void hide_popup_menu(void) {
+    if (D_8010D68E) {
+        switch (gPopupMenu->popupType) {
+            case POPUP_MENU_USE_ITEM:
+            case POPUP_MENU_SWITCH_PARTNER:
+            case POPUP_MENU_THROW_AWAY_ITEM:
+            case POPUP_MENU_TRADE_FOR_BADGE:
+            case POPUP_MENU_UPGRADE_PARTNER:
+            case POPUP_MENU_SELL_ITEM:
+            case POPUP_MENU_CHECK_ITEM:
+            case POPUP_MENU_CLAIM_ITEM:
+            case POPUP_MENU_READ_LETTER:
+            case POPUP_MENU_TAKE_FROM_CHEST:
+            case POPUP_MENU_USEKEY:
+                set_window_update(WINDOW_ID_14, (s32) basic_hidden_window_update);
+                set_window_update(WINDOW_ID_19, (s32) basic_hidden_window_update);
+                break;
+            case POPUP_MENU_READ_DIARY_PAGE:
+            case POPUP_MENU_READ_POSTCARD:
+            case POPUP_MENU_D:
+                set_window_update(WINDOW_ID_14, (s32) basic_hidden_window_update);
+                break;
+        }
+        set_window_update(WINDOW_ID_21, 2);
+    } else {
+        switch (gPopupMenu->popupType) {
+            case POPUP_MENU_USE_ITEM:
+            case POPUP_MENU_SWITCH_PARTNER:
+            case POPUP_MENU_THROW_AWAY_ITEM:
+            case POPUP_MENU_TRADE_FOR_BADGE:
+            case POPUP_MENU_UPGRADE_PARTNER:
+            case POPUP_MENU_SELL_ITEM:
+            case POPUP_MENU_CHECK_ITEM:
+            case POPUP_MENU_CLAIM_ITEM:
+            case POPUP_MENU_READ_LETTER:
+            case POPUP_MENU_TAKE_FROM_CHEST:
+            case POPUP_MENU_USEKEY:
+                set_window_update(WINDOW_ID_14, 2);
+                set_window_update(WINDOW_ID_19, 2);
+                break;
+            case POPUP_MENU_READ_DIARY_PAGE:
+            case POPUP_MENU_READ_POSTCARD:
+            case POPUP_MENU_D:
+                set_window_update(WINDOW_ID_14, 2);
+                break;
+        }
+    }
+
+    if (gPopupMenu->popupType == POPUP_MENU_TRADE_FOR_BADGE) {
+        set_window_update(WINDOW_ID_16, (s32) basic_hidden_window_update);
+        set_window_update(WINDOW_ID_20, (s32) basic_hidden_window_update);
+    }
+    if (gPopupMenu->popupType == POPUP_MENU_UPGRADE_PARTNER) {
+        set_window_update(WINDOW_ID_18, (s32) basic_hidden_window_update);
+    }
+    if (gPopupMenu->popupType == POPUP_MENU_SELL_ITEM) {
+        set_window_update(WINDOW_ID_16, (s32) basic_hidden_window_update);
+        set_window_update(WINDOW_ID_20, (s32) basic_hidden_window_update);
```

Comment:
We really need to retype this func to allow func ptrs as well as ints

Reply:
It's difficult, because doing so has messed with codegen unfortunately :/ but yeah we should revisit this for sure